uaclient/api/u/pro/security/vulnerabilities/_common/v1.py

```diff
@@ -0,0 +1,325 @@
+                    return False
+
+        return True
+
+    @staticmethod
+    def parse(manifest_file: str):
+        pkgs = {}
+
+        with open(manifest_file, "r") as f:
+            for line in f.readlines():
+                re_match = ProManifestSourcePackage.PKG_RE.match(line)
+                if re_match:
+                    match_groups = re_match.groupdict()
+                    pkg = match_groups["pkg"]
+
+                    if pkg == "snap":
+                        continue
+
+                    pkgs[pkg] = match_groups["version"]
+
+        return pkgs
+
+
+class SourcePackages:
+    SUPPORTED_MANIFESTS = [ProManifestSourcePackage]
+
+    def __init__(
+        self,
+        vulnerabilities_data: Dict[str, Any],
+        manifest_file: Optional[str] = None,
+    ):
+        self.manifest_file = manifest_file
+        self.vulnerabilities_data = vulnerabilities_data
+
+    def get(self):
+        if not self.manifest_file:
+            return query_installed_source_pkg_versions()
+        else:
+            return self.parse_manifest_file()
+
+    def parse_manifest_file(self) -> Dict[str, Dict[str, str]]:
+        if not self.manifest_file:
+            return {}
+
+        manifest_pkgs = None
+
+        for manifest_parser_cls in self.SUPPORTED_MANIFESTS:
+            if manifest_parser_cls.valid(self.manifest_file):
+                manifest_pkgs = manifest_parser_cls.parse(self.manifest_file)
+                break
+
+        if not manifest_pkgs:
+            raise UnsupportedManifestFile
+
+        source_pkgs = {}  # type: Dict[str, Dict[str, str]]
+        for pkg, version in manifest_pkgs.items():
+            source_pkg = _get_source_package_from_vulnerabilities_data(
+                self.vulnerabilities_data, pkg
+            )
+            if source_pkg in source_pkgs:
+                source_pkgs[source_pkg][pkg] = version
+            else:
+                source_pkgs[source_pkg] = {pkg: version}
+
+        return source_pkgs
+
+
+def _get_vulnerability_fix_status(
+    affected_packages: List[Dict[str, Optional[str]]],
+) -> VulnerabilityStatus:
+    vulnerability_status = VulnerabilityStatus.NO_FIX_AVAILABLE
+    num_fixes = 0
+    for pkg in affected_packages:
+        if pkg.get("fix_version") is not None:
+            num_fixes += 1
+
+    if num_fixes == len(affected_packages):
+        vulnerability_status = VulnerabilityStatus.FULL_FIX_AVAILABLE
+    elif num_fixes != 0:
+        vulnerability_status = VulnerabilityStatus.PARTIAL_FIX_AVAILABLE
+
+    return vulnerability_status
+
+
+class VulnerabilityParser:
+    vulnerability_type = None  # type: str
+
+    def __init__(
+        self,
+    ):
+        self.vulnerabilities = {}
+
+    def get_package_vulnerabilities(self, affected_pkg: Dict[str, Any]):
+        raise NotImplementedError
+
+    def parse_data(
+        self,
+        vulnerabilities_data: Dict[str, Any],
+        installed_pkgs_by_source: Dict[str, Dict[str, str]],
+    ):
```

Comment:
Consider looking at ways to make this long function a bit more readable for future us.